### ISSUE TITLE
Make Python object pointer accessible outside Variable class?

### ISSUE THREAD

**Issue by giacomofiorin** [lammps/lammps#438]
To use Python (embedded/callback mode) from the C++ code, the interpreter needs to be set up using `Py_Initialize()`.  But the initialization status is also used by the `Python` class to prevent issues between the "embedded" and the "library" modes: 
```
  if (pyMain == NULL) {
    if (Py_IsInitialized())
      error->all(FLERR,"Cannot embed Python when also "
                 "extending Python with LAMMPS");
    Py_Initialize();
    ...
  }
```
and `pyMain` is private in the `Python` class.  Furthermore, the pointer to the `Python` object is a private member of the `Variable` object, which could probably allow calling of Python code through `invoke_function()`, but certainly won't a return `PyObject *`, only a string.

Can the `Python` object be made available to other classes, or at least the pointer `pyMain` via some accessor?

**Comment by rbberger** [comment]
The `Py_IsInitialized` branch protects against a limitation of the current implementation of the LAMMPS Python binding. It doesn't consider threads. Trying to use it anyway will cause a nice segfault. The reason for this is that the current code doesn't honor the Python GIL (Global Interpreter Lock).

**Comment by rbberger** [comment]
My PR fixes the mixed use scenario. So you no longer get an error if you launch LAMMPS through Python and then try to execute Python code inside of LAMMPS.

You should be able to create your own instance of the Python class to do whatever you want. All instances uses the same global interpreter. (The init API calls are NOPs if called twice). If you absolutely need that `__main__` PyObject* you can always get it again with the same `PyImport_AddModule` call. From the [docs](https://docs.python.org/2.7/c-api/import.html#c.PyImport_AddModuleObject):

> Note: This function does not load or import the module; if the module wasn’t already loaded, you will get an empty module object.

**Comment by giacomofiorin** [comment]
Excellent, thank you!

What I'm trying to do is getting a pointer to the `PyObject` corresponding to a Python function given its name. To the best of my knowledge, you need the pointer to the corresponding module to use `PyObject_GetAttrString()` or another variant of the `getattr`call.  This is also what the `Python` class in LAMMPS does.

It is understood that the global interpreter lock should be used from other calls to Python (good to know that the interpreter is not exclusive to thread 0, btw).

Two questions:
1. In your PR, `Py_IsInitialized()` is still used to check whether the Python interpreter is calling LAMMPS as a a library.  I guess that no other code path should issue `Py_Initialize()`, and that the best way to ensure that the interpreter is ready is to issue a `python` command in the LAMMPS script early on.  Can you confirm?
2. If a separate `Python` object is created, would it share the database of functions with the object associated to the `python` script command?  My guess is not.

**Comment by akohlmey** [comment]
i think LAMMPS should support the option to have the same interpreter instance exposed to the different ways to hook python into LAMMPS, so that people can access the same functions and features. i would expect that people may be confused, if that doesn't happen, even if in some cases, complete encapsulation and thus no namespace clashes might be preferred.

i also think, that conceptually, the LAMMPS' Python wrapper class, is probably better instantiated in the LAMMPS class instead of Variable, which would also make it easier to provide access to other classes.

**Comment by rbberger** [comment]
@giacomofiorin:
1. `Py_Initialize()` is a no-op (does nothing) when it is called a second time in the same process. To ensure that the interpreter is initialized, currently you'll have to have at least one `python` command in your script.
2. If a second LAMMPS `Python` object is created it will use the same interpreter with the same module namespace. Any functions or variables defined up to that point will be available using the Python C API. What would be missing is the function mapping which is done in the LAMMPS `Python` object because that isn't shared between instances.

@akohlmey right now it is always the same interpreter and namespace on the Python side (unless `Py_Finalize` is called, e.g. by its destructor). You can have multiple interpreters, but that requires using what they call [Sub-Interpreters](https://docs.python.org/2/c-api/init.html#sub-interpreter-support). I agree it should be exposed higher up. Since there is no encapsulation right now anyway, it wouldn't change anything.

I'd suggest to have it globally available, but encapsulate it with a proxy class to allow lazy loading (only init the thing when there is something which needs it).

**Comment by giacomofiorin** [comment]
@rbberger I understand that `Py_Initialize()` can be called multiple times (with all calls but the first doing nothing.  But as you wrote also, you would need to have the `Python` object created and initialized for that to hold.  This means that outside of that object, `Py_Initialize()` can only called when it's a no-op.

I agree on making the `Python` object globally available (unfettered or through a proxy class) for as long as its initialization can be triggered from anywhere in the code.

**Comment by giacomofiorin** [comment]
The general interface is implemented by 2e1f8b4 and previous commits in @rbberger's Python refactoring PR.

I'm closing the issue, and will submit a PR with changes in the Colvars code that use the pointer to the `Python` object once I'm done testing it.
